### PR TITLE
New library for strings

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Separate a shared library from Audacity's source code tree.  But still to do -- correct the Linux build so that the library is installed in the correct folder.

This library may be re-used by the separate crash reporter executable.

The library contains string utilities, notably:

The classes Identifier and TaggedIdentifier<T> for wrapped string types that are used for internal purposes.

The class TranslatableString containing strings meant for the user to see.

Definitions of macros _, XO, etc. that mark strings for xgettext.

Namespace Languages which contains functions for initializing translation catalogs.

### CODE REVIEW

**Review by JamesCrook — APPROVED** [review]
lgtm

**Comment by Paul-Licameli** [comment]
Now I have fixed the Linux installation.  lib-strings.so goes to the same place as the wxWidgets libraries.

**Review by crsib — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -264,6 +264,14 @@ if( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
    set( CMAKE_MACOSX_RPATH OFF )
 endif()
 
+# Where Audacity libraries go
+if( CMAKE_SYSTEM_NAME MATCHES "Darwin|Windows" )
+   set( _LIBRARYDIR ${_EXEDIR} )
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply CMakeLists.txt:269]: See example packages here https://github.com/audacity/audacity/actions/runs/868681745

Yes there is inconsistency on MacOS (lib-strings.so is not in the same place as the wxWidgets libraries) but only that one.  I know it runs just fine.

File: CMakeLists.txt [hunk 2]
```diff
@@ -264,6 +264,14 @@ if( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
...
+else()
+   set( _LIBRARYDIR "${_DEST}/libraries" )
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply CMakeLists.txt:271]: What I wrote for Linux has the correct effect.   All the libraries under lib/audacity in the package

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply CMakeLists.txt:269]: Note that "modules" like mod-script-pipe are not in the same place as libraries nor need be.  They are in a special place examined for drop-ins.

File: libraries/CMakeLists.txt [hunk 3]
```diff
@@ -11,3 +12,10 @@ endforeach()
 
 set( GRAPH_EDGES "${GRAPH_EDGES}" PARENT_SCOPE )
 set( AUDACITY_LIBRARIES "${AUDACITY_LIBRARIES}" PARENT_SCOPE )
+
+if( NOT CMAKE_SYSTEM_NAME MATCHES "Darwin" )
+   if( NOT "${CMAKE_GENERATOR}" MATCHES "Visual Studio*")
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/CMakeLists.txt:17]: I was writing by analogy with some lines in modules/CMakeLists.txt which I didn't write

File: libraries/lib-strings/Internat.cpp [hunk 4]
```diff
@@ -47,7 +42,7 @@ wxArrayString Internat::exclude;
 // This function allows us to replace Audacity by DarkAudacity without peppering 
 // the source code with changes.  We split out this step, the customisation, as 
 // it is used on its own (without translation) in the wxTS macro.
-AUDACITY_DLL_API const wxString& GetCustomSubstitution(const wxString& str2)
+STRINGS_API const wxString& GetCustomSubstitution(const wxString& str2)
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-strings/Internat.cpp:45]: That was put there at 0885fda45971bab64f0600415daa435ec89d90b8 not by me, but I agree.

File: libraries/lib-strings/Languages.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+class wxArrayString;
+class wxString;
+
+#include "Identifier.h"
+#include "Internat.h"
+
+namespace Languages {
+
+STRINGS_API
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-strings/Languages.h:22]: I would want to check first that this does not confuse Doxygen when it generates HTML pages

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (crsib) [reply CMakeLists.txt:269]: >Yes there is inconsistency on MacOS (lib-strings.so is not in the same place as the wxWidgets libraries) but only that one. I know it runs just fine.

There will be many more. Both from the 3d party dependencies and our code. I would really love to be consistent here.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (crsib) [reply CMakeLists.txt:271]: Looks like they will go somewhere under libraries in this case

[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply CMakeLists.txt:271]: Observe where they really did go in artifacts:  https://github.com/audacity/audacity/actions/runs/868681745

This demonstrates that my solution works.

You don't see the complete solution here.  See the cooperating part in libraries/CMakeLists.txt where the linux-only install step moves the .so files from "${_DEST}/libraries into ${_PKGLIB} and it is that destination that matters.

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply CMakeLists.txt]: This commit is of course now superceded by your work, Dmitry, and is omitted from the latest rebasing.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply CMakeLists.txt]: mooted

File: libraries/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/CMakeLists.txt]: again never mind

File: libraries/lib-strings/Languages.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply libraries/lib-strings/Languages.h:22]: Experiment now shows, Doxygen works either way.  So, I will rewrite as you want it.

**Review by crsib — APPROVED** [review]

**Pull request merged**